extensions/csv/src/main/java/io/deephaven/csv/reading/CsvReader.java

```diff
@@ -0,0 +1,302 @@
+            if (firstDataRow == null) {
+                throw new RuntimeException(
+                        "Can't proceed because input file is empty and client has not specified headers");
+            }
+            headersToUse = new String[firstDataRow.length];
+            for (int ii = 0; ii < headersToUse.length; ++ii) {
+                headersToUse[ii] = "Column" + (ii + 1);
+            }
+        }
+
+        final int numCols = headersToUse.length;
+
+        // Create a DenseStorageWriter and two readers for each column.
+        final DenseStorageWriter[] dsws = new DenseStorageWriter[numCols];
+        final DenseStorageReader[] dsr0s = new DenseStorageReader[numCols];
+        final DenseStorageReader[] dsr1s = new DenseStorageReader[numCols];
+        for (int ii = 0; ii < numCols; ++ii) {
+            final DenseStorageWriter dsw = new DenseStorageWriter();
+            dsws[ii] = dsw;
+            dsr0s[ii] = dsw.newReader();
+            dsr1s[ii] = dsw.newReader();
+        }
+
+        // run in separate thread if async
+        // note: may need to have the writer in a special thread pool, because we don't want to ever starve the
+        // writer.
+        final String[] firstDataRowFinal = firstDataRow;
+        final ExecutorService exec = async ? Executors.newFixedThreadPool(4) : Executors.newSingleThreadExecutor();
+        final Future<Long> numRowsFuture = exec.submit(
+                () -> ParseInputFileToDenseStorage.doit(firstDataRowFinal, grabber, dsws));
+
+        final ArrayList<Future<Sink<?>>> sinkFutures = new ArrayList<>();
+
+        for (int ii = 0; ii < numCols; ++ii) {
+            final Set<ParserBase<?>> inferenceToUse = calcInferenceToUse(headersToUse[ii]);
+
+            final int iiCopy = ii;
+            final Future<Sink<?>> fcb = exec.submit(
+                    () -> ParseDenseStorageToColumn.doit(dsr0s[iiCopy], dsr1s[iiCopy],
+                            inferenceToUse, nullParser, nullContext, sinkFactory));
+            sinkFutures.add(fcb);
+        }
+
+        final long numRows;
+        Sink<?>[] sinks = new Sink<?>[numCols];
+        try {
+            numRows = numRowsFuture.get();
+            for (int ii = 0; ii < numCols; ++ii) {
+                sinks[ii] = sinkFutures.get(ii).get();
+            }
+        } catch (Exception ee) {
+            // TODO(kosak): not an IOException
+            throw new IOException("Caught exception while parsing columns", ee);
+        }
+
+        return new Result(numRows, headersToUse, sinks);
+    }
+
+    private Set<ParserBase<?>> calcInferenceToUse(String columnName) {
+        ParserBase<?> specifiedParser = parsers.get(columnName);
+        if (specifiedParser == null) {
+            return inference;
+        }
+        Set<ParserBase<?>> result = new HashSet<>();
+        result.add(specifiedParser);
+        return result;
+    }
+
+    private static String[] tryReadOneRow(CellGrabber grabber) throws IOException {
```

Comment:
final grabber?  In general, final in arguments? (seems like the trend follow in the methods below)

Reply:
Yeah I don't love "final" (especially for short methods) but it seems to be a Deephaven thing, so, when in Rome...